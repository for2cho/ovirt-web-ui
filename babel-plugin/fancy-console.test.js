import * as babel from '@babel/core'
import plugin from './fancy-console'

// remove indentation from a string in a way similar to how babel does
function deindent (strings, ...keys) {
  const out = []
  strings.forEach((part, index) => {
    out.push(part.replace(/\n +/g, '\n'))
    if (keys.length > 0) {
      out.push(keys.shift())
    }
  })
  out.unshift(out.shift().replace(/^ *\n */g, ''))
  out.push(out.pop().replace(/\n *$/g, ''))
  return out.join('')
}

const SRC_TESTS = [
  [
    'log',
    `
      const foo = 1;
      const bar = {};

      console.log("just a message");
      console.log(foo);
      console.log("foo", foo);
      console.log("foo", foo, "bar", bar);
      console.log("%cFANCY%c - ", "color: red", "", foo);
    `,
  ],
  [
    'info',
    `
      const foo = 1;
      const bar = {};

      console.info('just a message');
      console.info(bar);
      console.info('bar', bar);
      console.info(\`bar: \${bar}\`);
    `,
  ],
  [
    'warn',
    `
      const foo = 1;
      const bar = {};

      console.warn('just a message');
      console.warn(bar);
      console.warn('bar', bar);
      console.warn(\`bar: \${bar}\`);
    `,
  ],
  [
    'error',
    `
      const foo = 1;
      const bar = {};

      console.error('just a message');
      console.error(bar);
      console.error('bar', bar);
      console.error(\`bar: \${bar}\`);
    `,
  ],
  [
    'group',
    `
      console.group('groupName');
      console.group('groupName, level 2');
      console.groupEnd();
      console.groupEnd();
    `,
    deindent`
      console.group('groupName');
      console.group('groupName, level 2');
      console.groupEnd();
      console.groupEnd();
    `,
  ],
]

describe('fancy-console-test', () => {
  it('deindent', () => {
    const a = 'A'
    const b = 'B'

    const src = deindent`
      ${a}
      const foo;
      something();
      {${b}}
    `

    const res = `${a}
const foo;
something();
{${b}}`

    expect(src).toMatch(res)
  })

  test.each(SRC_TESTS)(
    'test [%s]',
    (title, src, expected = '') => {
      const { code } = babel.transformSync(src, { plugins: [plugin], babelrc: false })
      expect(code).toMatchSnapshot()
      if (expected.length > 0) {
        expect(code).toMatch(expected)
      }
    }
  )

  test('custom options', () => {
    const enhancements = {
      log: ['__ debug __'],
      info: ['^^ info ^^'],
      warn: ['** warn **'],
    }

    const src = `
      con.log("this is debug");
      con.info("this in info", "data");
      con.warn("and a warning %s here", "string");
      con.error("simple error");
      con.start();
      con.start("with a string argument");
      con.stop();
      con.stop("with a string argument");
    `

    const res = deindent`
      con.log("${enhancements.log[0]}" + (" " + "this is debug"));
      con.info("${enhancements.info[0]}" + (" " + "this in info"), "data");
      con.warn("${enhancements.warn[0]}" + (" " + "and a warning %s here"), "string");
      con.error("simple error");
      con.start();
      con.start("with a string argument");
      con.stop();
      con.stop("with a string argument");
    `

    const { code } = babel.transformSync(
      src,
      {
        plugins: [
          [
            plugin,
            {
              loggers: [
                {
                  object: 'con',
                  methods: Object.keys(enhancements),
                  enhancements,
                },
              ],
            },
          ],
        ],
        babelrc: false,
      })

    expect(code).toMatch(res)
  })
})
