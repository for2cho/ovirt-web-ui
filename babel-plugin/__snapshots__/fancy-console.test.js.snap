// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fancy-console-test test [error] 1`] = `
"const foo = 1;
const bar = {};
console.error(\\"%c error %c\\" + (\\" \\" + 'just a message'), \\"font-weight: bold; background-color: #ed403c; color: white;\\", \\"\\");
console.error(\\"%c error %c\\", \\"font-weight: bold; background-color: #ed403c; color: white;\\", \\"\\", bar);
console.error(\\"%c error %c\\" + (\\" \\" + 'bar'), \\"font-weight: bold; background-color: #ed403c; color: white;\\", \\"\\", bar);
console.error(\\"%c error %c\\" + (\\" \\" + \`bar: \${bar}\`), \\"font-weight: bold; background-color: #ed403c; color: white;\\", \\"\\");"
`;

exports[`fancy-console-test test [group] 1`] = `
"console.group('groupName');
console.group('groupName, level 2');
console.groupEnd();
console.groupEnd();"
`;

exports[`fancy-console-test test [info] 1`] = `
"const foo = 1;
const bar = {};
console.info(\\"%c info %c\\" + (\\" \\" + 'just a message'), \\"font-weight: bold; background-color: #01acac; color: white;\\", \\"\\");
console.info(\\"%c info %c\\", \\"font-weight: bold; background-color: #01acac; color: white;\\", \\"\\", bar);
console.info(\\"%c info %c\\" + (\\" \\" + 'bar'), \\"font-weight: bold; background-color: #01acac; color: white;\\", \\"\\", bar);
console.info(\\"%c info %c\\" + (\\" \\" + \`bar: \${bar}\`), \\"font-weight: bold; background-color: #01acac; color: white;\\", \\"\\");"
`;

exports[`fancy-console-test test [log] 1`] = `
"const foo = 1;
const bar = {};
console.log(\\"%c debug %c\\" + (\\" \\" + \\"just a message\\"), \\"font-weight: bold; background-color: #21409a; color: white;\\", \\"\\");
console.log(\\"%c debug %c\\", \\"font-weight: bold; background-color: #21409a; color: white;\\", \\"\\", foo);
console.log(\\"%c debug %c\\" + (\\" \\" + \\"foo\\"), \\"font-weight: bold; background-color: #21409a; color: white;\\", \\"\\", foo);
console.log(\\"%c debug %c\\" + (\\" \\" + \\"foo\\"), \\"font-weight: bold; background-color: #21409a; color: white;\\", \\"\\", foo, \\"bar\\", bar);
console.log(\\"%c debug %c\\" + (\\" \\" + \\"%cFANCY%c - \\"), \\"font-weight: bold; background-color: #21409a; color: white;\\", \\"\\", \\"color: red\\", \\"\\", foo);"
`;

exports[`fancy-console-test test [warn] 1`] = `
"const foo = 1;
const bar = {};
console.warn(\\"%c warn %c\\" + (\\" \\" + 'just a message'), \\"font-weight: bold; background-color: #f8a51b; color: white;\\", \\"\\");
console.warn(\\"%c warn %c\\", \\"font-weight: bold; background-color: #f8a51b; color: white;\\", \\"\\", bar);
console.warn(\\"%c warn %c\\" + (\\" \\" + 'bar'), \\"font-weight: bold; background-color: #f8a51b; color: white;\\", \\"\\", bar);
console.warn(\\"%c warn %c\\" + (\\" \\" + \`bar: \${bar}\`), \\"font-weight: bold; background-color: #f8a51b; color: white;\\", \\"\\");"
`;
